### PR TITLE
limit the number of active RPC HTTP connections

### PR DESCRIPTION
**Pull request by macfarla**
Signed-off-by: Sally MacFarlane <sally.macfarlane@consensys.net>

Added option to set an upper limit on RPC HTTP connections. Once this limit is reached, any further connections will be immediately closed. 

This will prevent Besu from crashing if the client keeps opening new connections, and does not close connections.

See #1687 

## Changelog

- [x] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

**Comment by RatanRSur** [comment]
So is the main value here to prevent the user from accidentally creating too many connections to the node? Because from the issue it seems like they were intentionally keeping them alive, which we should allow them to do since it's their node.

**Comment by macfarla** [comment]
> So is the main value here to prevent the user from accidentally creating too many connections to the node? Because from the issue it seems like they were intentionally keeping them alive, which we should allow them to do since it's their node.

This won't close any existing connections, just stop accepting new connections. So if they set keepalive that should still work. Basically this should stop Besu from crashing.

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/JsonRpcHttpService.java [hunk 1]
```diff
@@ -264,6 +275,34 @@ private void validateConfig(final JsonRpcConfiguration config) {
...
+    return connection -> {
+      if (activeConnectionsCount.get() >= maxActiveConnections) {
+        LOG.warn(
+            "Max Active Connections limit {} reached. Rejecting all new connections.",
+            maxActiveConnections);
+        // disallow new connections to prevent DoS
+        LOG.warn(
+            "Rejecting new client connection: remoteAddress: {} connection count: {}",
+            connection.remoteAddress(),
+            activeConnectionsCount.incrementAndGet());
```
Comment (lucassaldanha) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/JsonRpcHttpService.java:289]: We should probably log only one of these msgs. I believe we don't the two.

I think we can keep the msg when we reject the connection and change the msg slightly:
```
Rejected new HTTP connection from 1921.68.0.1. Max 80 active HTTP connections limit reached.
```
(or something similar)

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/JsonRpcHttpService.java [hunk 2]
```diff
@@ -264,6 +275,34 @@ private void validateConfig(final JsonRpcConfiguration config) {
...
+        connection.close();
+      } else {
+        LOG.warn(
```
Comment (lucassaldanha) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/JsonRpcHttpService.java:292]: This shouldn't be a warning. Change it to a debug message.

And we can change the msg slightly:
```
Opened HTTP connection from 192.160.1.1. Total of active HTTP connections: 12/80.
```

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/JsonRpcHttpService.java [hunk 3]
```diff
@@ -264,6 +275,34 @@ private void validateConfig(final JsonRpcConfiguration config) {
...
+            "New client connection: remoteAddress: {} connection count: {}",
+            connection.remoteAddress(),
+            activeConnectionsCount.incrementAndGet());
+      }
+      connection.closeHandler(
+          c ->
+              LOG.warn(
```
Comment (lucassaldanha) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/JsonRpcHttpService.java:299]: Not a warning. Change it to debug :)

And the msg:
```
Closed HTTP connection from 192.168.1.1. Total of active HTTP connection: 11/80
```

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/WebSocketService.java [hunk 4]
```diff
@@ -91,10 +91,8 @@ private WebSocketService(
                 new HttpServerOptions()
                     .setHost(configuration.getHost())
                     .setPort(configuration.getPort())
-                    .setHandle100ContinueAutomatically(true)
-                    .setCompressionSupported(true)
-                    .addWebSocketSubProtocol("undefined"))
-            .webSocketHandler(websocketHandler())
+                    .setWebsocketSubProtocols("undefined"))
```
Comment (lucassaldanha) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/WebSocketService.java:94]: What are these options that were removed?

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/LimitConnectionsJsonRpcHttpServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+import okhttp3.Response;
+import org.junit.Test;
+
+public class LimitConnectionsJsonRpcHttpServiceTest extends JsonRpcHttpServiceTestBase {
+
+  @Test
+  public void limitActiveConnections() throws Exception {
+    OkHttpClient newClient = new OkHttpClient();
+    int i;
+    for (i = 0; i < JsonRpcConfiguration.DEFAULT_MAX_ACTIVE_CONNECTIONS; i++) {
```
Comment (lucassaldanha) [inline ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/LimitConnectionsJsonRpcHttpServiceTest.java:31]: Is it possible to run this test with a lower limit? Probably 1? Then the second connection should fail.

It seems to be a waste of resources (and test time) opening 80 real connections to test the limit if we can reduce the config before starting the server.

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/WebSocketService.java (hunk 4 above)
Comment (macfarla) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/WebSocketService.java:94]: this was not meant to be changed...

Comment (macfarla) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/WebSocketService.java:94]: reverted

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/LimitConnectionsJsonRpcHttpServiceTest.java (hunk 5 above)
Comment (macfarla) [reply ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/LimitConnectionsJsonRpcHttpServiceTest.java]: refactored to add this

File: acceptance-tests/tests/src/test/java/org/hyperledger/besu/tests/acceptance/OpenTelemetryAcceptanceTest.java [hunk 6]
```diff
@@ -191,11 +191,11 @@ public void traceReporting() {
   @Test
   public void traceReportingWithTraceId() {
 
+    OkHttpClient okClient = new OkHttpClient();
```
Comment (macfarla) [inline acceptance-tests/tests/src/test/java/org/hyperledger/besu/tests/acceptance/OpenTelemetryAcceptanceTest.java:194]: moved this outside the wait so the connection gets reused

**Review by lucassaldanha — APPROVED** [review]

**Pull request merged**